In [24]:
import pandas as pd
import numpy as np

In [25]:
# marketing funnel
df_closed = pd.read_csv('cp1_dataset/Marketing_Funnel_Olist/olist_closed_deals_dataset.csv')
df_qualified = pd.read_csv('cp1_dataset/Marketing_Funnel_Olist/olist_marketing_qualified_leads_dataset.csv')

# customers
df_items = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/olist_order_items_dataset.csv')
df_payments = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/olist_order_payments_dataset.csv')
df_orders = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/olist_orders_dataset.csv')

df_cus = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/olist_customers_dataset.csv')
df_geo = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/olist_geolocation_dataset.csv')
df_products = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/olist_products_dataset.csv')
df_sellers = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/olist_sellers_dataset.csv')

df_translate = pd.read_csv('cp1_dataset/Brazilian_E_commerce_Olist/product_category_name_translation.csv')

In [98]:
df_qualified

mql_id first_contact_date  \
0     dac32acd4db4c29c230538b72f8dd87d         2018-02-01   
1     8c18d1de7f67e60dbd64e3c07d7e9d5d         2017-10-20   
2     b4bc852d233dfefc5131f593b538befa         2018-03-22   
3     6be030b81c75970747525b843c1ef4f8         2018-01-22   
4     5420aad7fec3549a85876ba1c529bd84         2018-02-21   
...                                ...                ...   
7995  feaba3ffcd2ff97501696c7f9a42f41c         2018-05-22   
7996  a79cb53cd009ab92e0143b92baa2407b         2018-03-27   
7997  68f049a23ab109c6a0f6989bb9a02994         2017-08-27   
7998  4f8c96e2509b984329044c6682c88ee9         2017-10-06   
7999  8a6492305a5fbcdcdd1a7f5a90764c07         2018-01-11   

                       landing_page_id          origin  
0     88740e65d5d6b056e0cda098e1ea6313          social  
1     007f9098284a86ee80ddeb25d53e0af8     paid_search  
2     a7982125ff7aa3b2054c6e44f9d28522  organic_search  
3     d45d558f0daeecf3cccdffe3c59684aa           email  
4     b48ec5f3b04e9068441002a19df93c6c  organic_search  
...                                ...             ...  
7995  e42a14209c69c3e9cc6b042620465f12     paid_search  
7996  c494978688ccf66ad9fad3d6a3338c22     paid_search  
7997  b48ec5f3b04e9068441002a19df93c6c  organic_search  
7998  a56671a54260a44923d32c2f08fad39c  organic_search  
7999  b76ef37428e6799c421989521c0e5077  direct_traffic  

[8000 rows x 4 columns]

- 유료 유입은 paid_search, social, display가 있다.

In [96]:
df_qualified.groupby('origin').count()

mql_id  first_contact_date  landing_page_id
origin                                                        
direct_traffic        499                 499              499
display               118                 118              118
email                 493                 493              493
organic_search       2296                2296             2296
other                 150                 150              150
other_publicities      65                  65               65
paid_search          1586                1586             1586
referral              284                 284              284
social               1350                1350             1350
unknown              1099                1099             1099

In [78]:
df_funnel = df_qualified.merge(df_closed, on='mql_id', how='inner')

mql_id first_contact_date  \
0    5420aad7fec3549a85876ba1c529bd84         2018-02-21   
1    a555fb36b9368110ede0f043dfc3b9a0         2018-04-04   
2    327174d3648a2d047e8940d7d15204ca         2018-04-03   
3    f5fee8f7da74f4887f5bcae2bafb6dd6         2018-01-14   
4    ffe640179b554e295c167a2f6be528e0         2017-10-09   
..                                ...                ...   
837  df9a2abe2ad3f2d8758b675ac8331ecf         2018-03-28   
838  8723201570415eb23b4f095d8c20a27c         2018-05-18   
839  2e57665b8faf05c967a801eb5aedfa0a         2018-05-25   
840  40955be51cc85c5a5a6cdd12b19e9f10         2018-02-04   
841  8a6492305a5fbcdcdd1a7f5a90764c07         2018-01-11   

                      landing_page_id          origin  \
0    b48ec5f3b04e9068441002a19df93c6c  organic_search   
1    2f08471dfb83baef522f952438a36f6e        referral   
2    1ceb590cd1e00c7ee95220971f82693d  organic_search   
3    22c29808c4f815213303f8933030604c     paid_search   
4    a0fbc0263cb98bdb0d72b9f9f283a172         unknown   
..                                ...             ...   
837  a7982125ff7aa3b2054c6e44f9d28522  organic_search   
838  b76ef37428e6799c421989521c0e5077  organic_search   
839  40dec9f3d5259a3d2dbcdab2114fae47     paid_search   
840  88740e65d5d6b056e0cda098e1ea6313          social   
841  b76ef37428e6799c421989521c0e5077  direct_traffic   

                            seller_id                            sdr_id  \
0    2c43fb513632d29b3b58df74816f1b06  a8387c01a09e99ce014107505b92388c   
1    bbb7d7893a450660432ea6652310ebb7  09285259593c61296eef10c734121d5b   
2    612170e34b97004b3ba37eae81836b4c  b90f87164b5f8c2cfa5c8572834dbe3f   
3    21e1781e36faf92725dde4730a88ca0f  56bf83c4bb35763a51c2baab501b4c67   
4    ed8cb7b190ceb6067227478e48cf8dde  4b339f9567d060bcea4f5136b9f5949e   
..                                ...                               ...   
837  bfcc27719640628da877db48b672b169  45749fb708130f78d0db07d8d80f030b   
838  4a82eab98441aeb64566e2776c1fb2b6  370c9f455f93a9a96cbe9bea48e70033   
839  55031883943971ca22db6894574cfe2c  370c9f455f93a9a96cbe9bea48e70033   
840  c0e933c238e41f0cd459d6025ee9b364  068066e24f0c643eb1d089c7dd20cd73   
841  1bba9c94eb5c4e82fba9f7286fe81025  56bf83c4bb35763a51c2baab501b4c67   

                                sr_id             won_date  \
0    4ef15afb4b2723d8f3d81e51ec7afefe  2018-02-26 19:58:54   
1    d3d1e91a157ea7f90548eef82f1955e3  2018-05-08 20:17:59   
2    6565aa9ce3178a5caf6171827af3a9ba  2018-06-05 17:27:23   
3    d3d1e91a157ea7f90548eef82f1955e3  2018-01-17 13:51:03   
4    d3d1e91a157ea7f90548eef82f1955e3  2018-07-03 20:17:45   
..                                ...                  ...   
837  a8387c01a09e99ce014107505b92388c  2018-08-01 20:22:31   
838  4b339f9567d060bcea4f5136b9f5949e  2018-09-27 18:58:41   
839  a8387c01a09e99ce014107505b92388c  2018-06-06 20:15:26   
840  d3d1e91a157ea7f90548eef82f1955e3  2018-03-06 15:52:23   
841  de63de0d10a6012430098db33c679b0b  2018-05-02 13:31:32   

                    business_segment        lead_type lead_behaviour_profile  \
0                                pet    online_medium                    cat   
1                    car_accessories         industry                  eagle   
2                    home_appliances       online_big                    cat   
3                         food_drink     online_small                    NaN   
4                    home_appliances         industry                   wolf   
..                               ...              ...                    ...   
837                       home_decor          offline                    cat   
838  construction_tools_house_garden       online_big                   wolf   
839                        computers    online_medium                   wolf   
840                              pet  online_beginner                    cat   
841              household_utilities         industry                  eagle   

    has_company ha

### 데이터 병합
- mql_id 기준으로 marketing funnel 먼저

In [ ]:
# marketing funnel 데이터 병합
df_funnel = df_qualified.merge(df_closed, on='mql_id', how='inner')

# df_funnel과 df_sellers 병합
df_funnel_sellers = df_funnel.merge(df_sellers, on='seller_id', how='inner')

# df_items 데이터 셋과 합치기
df_funnel_sellers_items = df_funnel_sellers.merge(df_items, on='seller_id', how='left')

df_funnel_sellers_items

In [ ]:
df_funnel_sellers_items.isna().sum()

In [ ]:
# 최근 활동 날짜를 구해주기 위해서 df_orders 데이터를 합쳐주자.
# df_orders에서 order_approved_at 컬럼 활용. (결제 승인 날짜 컬럼)
df_funnel_sellers_items_order = df_funnel_sellers_items.merge(df_orders, on='order_id', how='left')
df_funnel_sellers_items_order

In [29]:
df_funnel_sellers_items_order.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5044 entries, 0 to 5043
Data columns (total 33 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   mql_id                         5044 non-null   object 
 1   first_contact_date             5044 non-null   object 
 2   landing_page_id                5044 non-null   object 
 3   origin                         5033 non-null   object 
 4   seller_id                      5044 non-null   object 
 5   sdr_id                         5044 non-null   object 
 6   sr_id                          5044 non-null   object 
 7   won_date                       5044 non-null   object 
 8   business_segment               5044 non-null   object 
 9   lead_type                      4996 non-null   object 
 10  lead_behaviour_profile         3303 non-null   object 
 11  has_company                    30 non-null     object 
 12  has_gtin                       30 non-null     o

In [30]:
df_total = df_funnel_sellers_items_order.copy()

In [31]:
df_total['temp_LTV'] = df_total['price'] - df_total['freight_value']

In [32]:
df_total.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5044 entries, 0 to 5043
Data columns (total 34 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   mql_id                         5044 non-null   object 
 1   first_contact_date             5044 non-null   object 
 2   landing_page_id                5044 non-null   object 
 3   origin                         5033 non-null   object 
 4   seller_id                      5044 non-null   object 
 5   sdr_id                         5044 non-null   object 
 6   sr_id                          5044 non-null   object 
 7   won_date                       5044 non-null   object 
 8   business_segment               5044 non-null   object 
 9   lead_type                      4996 non-null   object 
 10  lead_behaviour_profile         3303 non-null   object 
 11  has_company                    30 non-null     object 
 12  has_gtin                       30 non-null     o

In [33]:
df_GB = df_total.groupby('seller_id')['temp_LTV'].sum()

In [101]:
df_GB

seller_id
01266d4c46afa519678d16a8b683d325      43.02
01fd077212124329bac32490e8ef80d9     632.64
02f623a8eb246f3c5f7c2f96462654e6    5548.37
03a2a6dcbe45d639c4dc2a0d24769c40     906.16
044668ccd5316b12a7cf0d54a156e3e9      84.25
                                     ...   
fd435faa3c0422b60440ea3480d0e77c     615.13
fe87f472055fbcf1d7e691c00b1560dc     955.64
ffa6adafb71b807dc13159e26431354c    2132.82
ffad1e7127fb622cb64a900751590acd     677.77
ffc470761de7d0232558ba5e786e57b7    1210.89
Name: temp_LTV, Length: 380, dtype: float64

### 병합 결과
- seller_id로 group by 했을 때, 데이터의 양이 많이 적어진다.
- 통상적으로 머신러닝 모델을 돌리기 위해서 최소 1000개~10000개 까지는 데이터가 확보되어야 하는데, 380개는 너무 적다.
- 이후 data augmentaion을 찾아보고, 이를 활용해보는 방법도 생각해보자.

 :https://velog.io/@ysn003/%EC%8B%9C%EA%B3%84%EC%97%B4-%EB%8D%B0%EC%9D%B4%ED%84%B0-Augmentation-%EB%B0%A9%EB%B2%95-%EC%A0%95%EB%A6%AC
- 하지만, 분석과 인사이트 도출에 초점을 맞춘다면, 데이터가 크지 않더라도, 그 데이터에서 도출해내는 것이 분석가의 역량이기 때문에, 패턴과 인사이트를 최대한 찾아내보자.
- 실제로 비즈니스에서도, 생각해보면 회사 과제들에서도 10000개 이상의 데이터를 준 적은 거의 없다. 분석가로써 할 수 있는 분석을 최대한 해보자.
- 참고 사이트

 :https://brunch.co.kr/@8d1b089f514b4d5/21
 :http://taewan.kim/tutorial_manual/ml_yearning/020.setup_dev_test_set/7/
- 가설 검정 및 통계를 활용해서 분석해보자


### 1. 먼저 할 수 있는 것 중, 판매자의 리텐션을 확인해보자.
- order_approved_at과 won_date의 차이로 리텐션을 먼저 구해보자

In [51]:
df_GB__ = df_total.groupby(['seller_id','origin','business_segment']).agg({'temp_LTV':'sum', 'seller_id':'count', 'won_date':'min', 'order_approved_at':'max'})

In [58]:
df_GB__.rename(columns = {'seller_id':'num_selles'}, inplace = True)

In [59]:
df_GB__.reset_index(drop = False, inplace = True)

In [60]:
df_GB__

seller_id  temp_LTV  num_selles  \
0    01266d4c46afa519678d16a8b683d325     43.02           3   
1    01fd077212124329bac32490e8ef80d9    632.64          39   
2    02f623a8eb246f3c5f7c2f96462654e6   5548.37          44   
3    03a2a6dcbe45d639c4dc2a0d24769c40    906.16           9   
4    044668ccd5316b12a7cf0d54a156e3e9     84.25           4   
..                                ...       ...         ...   
375  fd435faa3c0422b60440ea3480d0e77c    615.13           1   
376  fe87f472055fbcf1d7e691c00b1560dc    955.64           6   
377  ffa6adafb71b807dc13159e26431354c   2132.82          18   
378  ffad1e7127fb622cb64a900751590acd    677.77          15   
379  ffc470761de7d0232558ba5e786e57b7   1210.89          31   

                won_date    order_approved_at  
0    2018-05-18 20:02:04  2018-08-21 04:55:37  
1    2018-03-28 22:45:45  2018-08-25 04:05:18  
2    2018-04-18 20:16:19  2018-08-23 00:05:16  
3    2018-02-09 18:11:10  2018-08-15 18:35:13  
4    2018-01-31 21:00:13  2018-08-20 21:30:01  
..                   ...                  ...  
375  2018-07-26 17:03:01  2018-08-20 14:11:42  
376  2018-01-23 12:41:16  2018-08-20 13:30:46  
377  2018-02-20 17:44:19  2018-08-14 04:25:23  
378  2018-02-02 20:42:47  2018-07-14 18:04:15  
379  2018-02-14 11:48:29  2018-08-21 04:29:47  

[380 rows x 5 columns]

In [64]:
# Retention 계산
# order_approved_at와 won_date 이 str형태이기 때문에, to_datetime 적용먼저
df_GB__['order_approved_at'] = pd.to_datetime(df_GB__['order_approved_at'])
df_GB__['won_date'] = pd.to_datetime(df_GB__['won_date'])

# 이제 Retention 계산
df_GB__['retention'] = df_GB__['order_approved_at'] - df_GB__['won_date']

In [65]:
df_GB__

seller_id  temp_LTV  num_selles  \
0    01266d4c46afa519678d16a8b683d325     43.02           3   
1    01fd077212124329bac32490e8ef80d9    632.64          39   
2    02f623a8eb246f3c5f7c2f96462654e6   5548.37          44   
3    03a2a6dcbe45d639c4dc2a0d24769c40    906.16           9   
4    044668ccd5316b12a7cf0d54a156e3e9     84.25           4   
..                                ...       ...         ...   
375  fd435faa3c0422b60440ea3480d0e77c    615.13           1   
376  fe87f472055fbcf1d7e691c00b1560dc    955.64           6   
377  ffa6adafb71b807dc13159e26431354c   2132.82          18   
378  ffad1e7127fb622cb64a900751590acd    677.77          15   
379  ffc470761de7d0232558ba5e786e57b7   1210.89          31   

               won_date   order_approved_at         retention  
0   2018-05-18 20:02:04 2018-08-21 04:55:37  94 days 08:53:33  
1   2018-03-28 22:45:45 2018-08-25 04:05:18 149 days 05:19:33  
2   2018-04-18 20:16:19 2018-08-23 00:05:16 126 days 03:48:57  
3   2018-02-09 18:11:10 2018-08-15 18:35:13 187 days 00:24:03  
4   2018-01-31 21:00:13 2018-08-20 21:30:01 201 days 00:29:48  
..                  ...                 ...               ...  
375 2018-07-26 17:03:01 2018-08-20 14:11:42  24 days 21:08:41  
376 2018-01-23 12:41:16 2018-08-20 13:30:46 209 days 00:49:30  
377 2018-02-20 17:44:19 2018-08-14 04:25:23 174 days 10:41:04  
378 2018-02-02 20:42:47 2018-07-14 18:04:15 161 days 21:21:28  
379 2018-02-14 11:48:29 2018-08-21 04:29:47 187 days 16:41:18  

[380 rows x 6 columns]

In [66]:
df_GB__['retention'].describe()

count                            380
mean     115 days 12:18:04.931578948
std       50 days 18:29:45.484489947
min                  7 days 07:18:08
25%          82 days 20:37:21.500000
50%                112 days 09:39:41
75%         153 days 07:25:08.250000
max                220 days 22:18:54
Name: retention, dtype: object